In [62]:
import pandas as pd


In [63]:
#Loading Data Sets
Ahmedabad_data = pd.read_csv('Ahmedabad.csv')
Banglore_data = pd.read_csv('Banglore.csv')
Car_data = pd.read_csv('Car_data.csv')
Chennai_data = pd.read_csv('Chennai.csv')
Gurgaon_data = pd.read_csv('Gurgaon.csv')
Hyderabad_data = pd.read_csv('Hyderabad.csv')
Jaipur_data = pd.read_csv('Jaipur.csv')
Kolkata_data = pd.read_csv('Kolkata.csv')
Mumbai_data = pd.read_csv('Mumbai.csv')
Pune_data = pd.read_csv('Pune.csv')

Data Preprocessing

In [64]:
#combining all the data sets
combined_data = pd.concat([Ahmedabad_data, Banglore_data, Car_data, Chennai_data, Gurgaon_data, Hyderabad_data, Jaipur_data, Kolkata_data, Mumbai_data, Pune_data], ignore_index=True)



In [65]:
Data = combined_data


In [80]:
Data.head()


Unnamed: 0                    Brand  Make_Year    Fuel  KMs_Driven  \
0           0  2020 Maruti Swift Dzire       2019  Petrol  11,935 Kms   
1           1      2016 Nissan Terrano       2016  Diesel  74,555 Kms   
2           2  2020 Maruti Swift Dzire       2019  Petrol   9,379 Kms   
3           3        2018 Maruti Swift       2018  Diesel  31,824 Kms   
4           4       2017 Maruti Baleno       2017  Petrol  49,072 Kms   

  Engine_Displacement No_Of_Owner Transmission     Mileage Max_Power  ...  \
0             1197 cc   1st Owner       Manual  21.21 kmpl  81.80bhp  ...   
1             1461 cc   2nd Owner       Manual  20.45 kmpl   83.8bhp  ...   
2             1197 cc   1st Owner       Manual  21.21 kmpl  81.80bhp  ...   
3             1248 cc   1st Owner       Manual   28.4 kmpl     74bhp  ...   
4             1197 cc   1st Owner       Manual  21.01 kmpl  81.80bhp  ...   

  Seats  Color Gear_Box Drive_Type Steering_Type Front_Brake_Type  \
0     5   Grey  5 Speed        FWD         Power             Disc   
1     5  Black  5 Speed        FWD         Power             Disc   
2     5  White  5 Speed        FWD         Power             Disc   
3     5    Red        5        FWD         Power             Disc   
4     5   Blue  5 Speed        FWD      Electric             Disc   

  Rear_Brake_Type   Acceleration          Engine_Type     Price  
0            Drum   12.6 Seconds  K Series VVT Engine  6,80,000  
1            Drum     14 Seconds    K9K Diesel Engine  4,70,500  
2            Drum   12.6 Seconds  K Series VVT Engine  6,75,000  
3            Drum            NaN      DDiS 190 Engine  6,00,000  
4            Drum  12.36 seconds      1.2L VVT Engine  6,10,000  

[5 rows x 21 columns]

Inspecting the Data: Check data types, null values, and the general structure of the dataset.

In [81]:
Data.info()  # Shows data types and missing values



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10394 entries, 0 to 10393
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Unnamed: 0           10394 non-null  int64 
 1   Brand                10394 non-null  object
 2   Make_Year            10394 non-null  int64 
 3   Fuel                 10394 non-null  object
 4   KMs_Driven           10394 non-null  object
 5   Engine_Displacement  10394 non-null  object
 6   No_Of_Owner          10394 non-null  object
 7   Transmission         10394 non-null  object
 8   Mileage              10394 non-null  object
 9   Max_Power            10394 non-null  object
 10  Torque               10394 non-null  object
 11  Seats                10394 non-null  int64 
 12  Color                10394 non-null  object
 13  Gear_Box             10386 non-null  object
 14  Drive_Type           10208 non-null  object
 15  Steering_Type        10344 non-null  object
 16  Fron

**DataFrame Summary**

Total Entries: 10,394 rows (indexed from 0 to 10,393).

Total Columns: 21 columns.

**Column Details**

**Data Types:**

int64: 3 columns (e.g., Unnamed: 0, Make_Year, Seats).

object: 18 columns (e.g., Brand, Fuel, KMs_Driven, etc.).

**Non-Null Counts**

Most columns have 10,394 non-null entries, indicating no missing values.

Some columns have missing values:

Gear_Box: 10,386 non-null entries.

Drive_Type: 10,208 non-null entries.

Steering_Type: 10,344 non-null entries.

Front_Brake_Type: 10,346 non-null entries.

Rear_Brake_Type: 10,346 non-null entries.

Acceleration: 9,320 non-null entries.

Engine_Type: 5,144 non-null entries.

**Memory Usage**

The DataFrame uses approximately 1.7 MB of memory.

**Conclusion**
The DataFrame contains a mix of integer and object data types, with some columns having missing values. Most columns are fully populated, while a few require attention for missing data.

In [82]:
Data.describe()  # Summary statistics


Unnamed: 0     Make_Year         Seats
count  10394.000000  10394.000000  10394.000000
mean    1460.271984   2014.891861      5.188378
std     1563.455786      3.094559      0.653699
min        0.000000   1995.000000      2.000000
25%      262.000000   2013.000000      5.000000
50%      594.000000   2015.000000      5.000000
75%     2597.750000   2017.000000      5.000000
max     5196.000000   2021.000000      9.000000

In [68]:
print(Data.head())  # First few rows for a quick look

   Unnamed: 0                    Brand  Make_Year    Fuel  KMs_Driven  \
0           0  2020 Maruti Swift Dzire       2019  Petrol  11,935 Kms   
1           1      2016 Nissan Terrano       2016  Diesel  74,555 Kms   
2           2  2020 Maruti Swift Dzire       2019  Petrol   9,379 Kms   
3           3        2018 Maruti Swift       2018  Diesel  31,824 Kms   
4           4       2017 Maruti Baleno       2017  Petrol  49,072 Kms   

  Engine_Displacement No_Of_Owner Transmission     Mileage Max_Power  ...  \
0             1197 cc   1st Owner       Manual  21.21 kmpl  81.80bhp  ...   
1             1461 cc   2nd Owner       Manual  20.45 kmpl   83.8bhp  ...   
2             1197 cc   1st Owner       Manual  21.21 kmpl  81.80bhp  ...   
3             1248 cc   1st Owner       Manual   28.4 kmpl     74bhp  ...   
4             1197 cc   1st Owner       Manual  21.01 kmpl  81.80bhp  ...   

  Seats  Color Gear_Box Drive_Type Steering_Type Front_Brake_Type  \
0     5   Grey  5 Speed      

In [83]:
#Data Cleaning
# Get the count of missing values in each column
missing_values = Data.isnull().sum()
print(missing_values)



Unnamed: 0                0
Brand                     0
Make_Year                 0
Fuel                      0
KMs_Driven                0
Engine_Displacement       0
No_Of_Owner               0
Transmission              0
Mileage                   0
Max_Power                 0
Torque                    0
Seats                     0
Color                     0
Gear_Box                  8
Drive_Type              186
Steering_Type            50
Front_Brake_Type         48
Rear_Brake_Type          48
Acceleration           1074
Engine_Type            5250
Price                     0
dtype: int64


From above we can see that the missing values are in:

Drive_Type  
Steering_Type  
Front_Brake_Type  
Rear_Brake_Type  
Acceleration and 
Engine_Type 

In [84]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

In [95]:
# 1. Handle Missing Values
# Fill missing values for categorical columns with the mode
for column in Data.select_dtypes(include=['object']).columns:
    Data[column].fillna(Data[column].mode()[0], inplace=True)
    
    # Fill missing values for numerical columns with the mean
    for column in Data.select_dtypes(include=['int64', 'float64']).columns:
        Data[column].fillna(Data[column].mean(), inplace=True)


In [94]:
#Checking for missing values after handling
missing_values = Data.isnull().sum()
print(missing_values)



Unnamed: 0             0
Brand                  0
Make_Year              0
Fuel                   0
KMs_Driven             0
Engine_Displacement    0
No_Of_Owner            0
Transmission           0
Mileage                0
Max_Power              0
Torque                 0
Seats                  0
Color                  0
Gear_Box               0
Drive_Type             0
Steering_Type          0
Front_Brake_Type       0
Rear_Brake_Type        0
Acceleration           0
Engine_Type            0
Price                  0
dtype: int64


In [96]:
# 2. Convert Data Types
Data['Make_Year'] = Data['Make_Year'].astype(int)  # Ensure Make_Year is int
Data['Seats'] = Data['Seats'].astype(int)  # Ensure Seats is int


In [ ]:
# 3. Remove Duplicates
Data.drop_duplicates(inplace=True)